### PR TITLE
print as function for Python 3 transition

### PR DESCRIPTION
**Pull request by smoe**
In analogy to https://github.com/miRTop/mirtop/pull/45 for the dev branch.

### CODE REVIEW

File: mirtop/exporter/isomirs.py [hunk 1]
```diff
@@ -31,10 +32,10 @@ def _read_file(fn, precursors, matures, out_dir):
     samples = read_samples(fn)
     for sample in samples:
         with open(os.path.join(out_dir, "%s.mirna" % sample), 'w') as outh:
-            print >>outh, "\t".join(
+            print("\t".join(
                 ["seq", "name", "freq", "mir", "start", "end",
                  "mism", "add", "t5", "t3", "s5", "s3", "DB",
-                 "precursor", "ambiguity"])
+                 "precursor", "ambiguity"]), file=outh, end="")
```
Comment (lpantano) [inline mirtop/exporter/isomirs.py]: This need to be with return carrier at the end, maybe `end="\n"`?

File: mirtop/exporter/isomirs.py [hunk 2]
```diff
@@ -72,4 +73,4 @@ def _read_file(fn, precursors, matures, out_dir):
                 with open(os.path.join(out_dir, "%s.mirna" % sample),
                           'a') as outh:
                     line[2] = counts
-                    print >>outh, "\t".join(line)
+                    print("\t".join(line), file=outh, end="")
```
Comment (lpantano) [inline mirtop/exporter/isomirs.py]: `end="\n"`

File: mirtop/gff/__init__.py [hunk 3]
```diff
@@ -57,11 +58,11 @@ def reader(args):
 
 def _write(lines, header, fn):
     out_handle = open(fn, 'w')
-    print >>out_handle, header
+    print(header, file=out_handle, end="")
     for m in lines:
         for s in sorted(lines[m].keys()):
             for hit in lines[m][s]:
-                print >>out_handle, hit[4]
+                print(hit[4], file=out_handle, end="")
```
Comment (lpantano) [inline mirtop/gff/__init__.py]: `end="\n"`

File: mirtop/gff/__init__.py [hunk 4]
```diff
@@ -57,11 +58,11 @@ def reader(args):
 
 def _write(lines, header, fn):
     out_handle = open(fn, 'w')
-    print >>out_handle, header
+    print(header, file=out_handle, end="")
```
Comment (lpantano) [inline mirtop/gff/__init__.py]: `end="\n"`

File: mirtop/gff/compare.py [hunk 5]
```diff
@@ -34,11 +36,11 @@ def compare(args):
         fn_out = os.path.join(args.out, "summary.txt")
         with open(fn_out, 'w') as outh:
             for fn in result:
-                print >>outh,  "sample\tidu\tseq\ttag\tsame_mirna\t%s" % "\t".join(result[fn][0][3].keys())
+                print("sample\tidu\tseq\ttag\tsame_mirna\t%s" % "\t".join(result[fn][0][3].keys()), file=outh, end="")
```
Comment (lpantano) [inline mirtop/gff/compare.py]: `end="\n"`

File: mirtop/gff/compare.py [hunk 6]
```diff
@@ -34,11 +36,11 @@ def compare(args):
...
                 for line in result[fn]:
                     read = read_id(line[0])
                     acc = "\t".join([line[3][v] for v in line[3]])
-                    print >>outh, "%s\t%s\t%s\t%s\t%s\t%s" % (fn, line[0], read, line[1], line[2], acc)
+                    print("%s\t%s\t%s\t%s\t%s\t%s" % (fn, line[0], read, line[1], line[2], acc), file=outh, end="")
```
Comment (lpantano) [inline mirtop/gff/compare.py]: `end="\n"`

File: mirtop/gff/convert.py [hunk 7]
```diff
@@ -46,7 +48,7 @@ def convert_gff_counts(args):
                 samples = sep.join(samples_line.strip().split("COLDATA:")[1].strip().split(","))
                 header = sep.join(['UID', 'Read', 'miRNA', 'Variant',
                                    variant_header, samples])
-                print >>outh, header
+                print(header, file=outh, end="")
```
Comment (lpantano) [inline mirtop/gff/convert.py]: `end="\n"`

File: mirtop/gff/convert.py [hunk 8]
```diff
@@ -81,7 +83,7 @@ def convert_gff_counts(args):
             summary = sep.join([UID, Read,  mirna, variant,
                                 cols_variants, expression])
             logger.debug(summary)
-            print >>outh, summary
+            print(summary, file=outh, end="")
```
Comment (lpantano) [inline mirtop/gff/convert.py]: `end="\n"`

File: mirtop/gff/stats.py [hunk 9]
```diff
@@ -34,12 +36,12 @@ def stats(args):
     outfn = os.path.join(args.out, "mirtop_stats.txt")
     if args.out != "tmp_mirtop":
         with open(outfn, 'w') as outh:
-            print >>outh, message_info
+            print(message_info, file=outh, end="")
```
Comment (lpantano) [inline mirtop/gff/stats.py]: `end="\n"`

File: mirtop/libs/simulator.py [hunk 10]
```diff
@@ -77,14 +78,14 @@ def _write_reads(reads, prefix):
     out_fasta = prefix + ".fasta"
     out_real = prefix + ".txt"
     with open(out_ma, 'w') as ma_handle:
-        print >>ma_handle, "id\tseq\tsample"
+        print("id\tseq\tsample", file=ma_handle, end="")
```
Comment (lpantano) [inline mirtop/libs/simulator.py]: `end="\n"`

File: mirtop/libs/simulator.py [hunk 11]
```diff
@@ -77,14 +78,14 @@ def _write_reads(reads, prefix):
...
         with open(out_fasta, 'w') as fa_handle:
             with open(out_real, 'w') as read_handle:
                 for idx, r in enumerate(reads):
                     info = r.split("_")
-                    print >>ma_handle, "seq_%s\t%s\t%s" % (idx, reads[r][0], reads[r][1])
-                    print >>fa_handle, ">seq_%s\n%s" % (idx, reads[r][0])
-                    print >>read_handle, "%s\t%s\t%s\t%s\t%s\t%s\t%s" % (idx, r, reads[r][0], reads[r][1], info[1], info[2], info[3])
+                    print("seq_%s\t%s\t%s" % (idx, reads[r][0], reads[r][1]), file=ma_handle, end="")
+                    print(">seq_%s\n%s" % (idx, reads[r][0]), file=fa_handle, end="")
+                    print("%s\t%s\t%s\t%s\t%s\t%s\t%s" % (idx, r, reads[r][0], reads[r][1], info[1], info[2], info[3]), file=read_handle, end="")
```
Comment (lpantano) [inline mirtop/libs/simulator.py]: `end="\n"` in all previous line as well

File: mirtop/mirna/snps.py [hunk 12]
```diff
@@ -87,7 +89,7 @@ def print_vcf(data):
     info = "ID=%s" % data['mature']
     frmt = "GT:NR:NS"
     gntp = "%s:%s:%s" % (_genotype(data), data["counts"], data["diff"])
-    print >>STDOUT, "\t".join(map(str, [chrom, pos, id_name, nt_ref, nt_snp, qual, flt, info, frmt, gntp]))
+    print("\t".join(map(str, [chrom, pos, id_name, nt_ref, nt_snp, qual, flt, info, frmt, gntp])),file=STDOUT,end="")
```
Comment (lpantano) [inline mirtop/mirna/snps.py]: `end="\n"` for all print on file in this file

File: scripts/miRNA.simulator.py [hunk 13]
```diff
@@ -120,11 +122,11 @@ def create_iso(name, mir, seq, numsim, exp):
 
 def _write(lines, header, fn):
     out_handle = open(fn, 'w')
-    print >>out_handle, header
+    print(header, file=out_handle, end="")
     for m in lines:
         for s in sorted(lines[m].keys()):
             for hit in lines[m][s]:
-                print >>out_handle, hit[4]
+                print(hit[4], file=out_handle, end="")
```
Comment (lpantano) [inline scripts/miRNA.simulator.py]: `end="\n"` for all print on file in this file

File: mirtop/mirna/snps.py (hunk 12 above)
Comment (lpantano) [reply mirtop/mirna/snps.py]: sorry, in this case for all print on STDOUT

**Comment by smoe** [comment]
I should have tried that myself before submitting.
The explicit "\n" may (or may not) bring trouble when mixing Windows, Mac and Linux worlds?

I just empirically found out that the information on the end being required that way was either missunderstood on my side or wrong. The default behaviour was just fine. The end="" shall go, this means. I think I apologised before. I do it again: sorry.

**Comment by lpantano** [comment]
awesome! you don’t need to apologize! thanks for the time you are investing, really appreciate it! I’ll merge it tomorrow when I habe my computer again.

**Pull request merged**